### ISSUE TITLE
test_xmlrpc leaks dangling threads

### ISSUE THREAD

**Issue by vstinner** [python/cpython#75430]
BPO | [31247](https://bugs.python.org/issue31247)
--- | :---
Nosy | @vstinner
PRs | <li>python/cpython#3166</li><li>python/cpython#3168</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-08-21.16:39:53.521>
created_at = <Date 2017-08-21.15:48:53.393>
labels = ['3.7', 'tests', 'performance']
title = 'test_xmlrpc leaks dangling threads'
updated_at = <Date 2017-08-21.16:39:53.519>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-08-21.16:39:53.519>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-08-21.16:39:53.521>
closer = 'vstinner'
components = ['Tests']
creation = <Date 2017-08-21.15:48:53.393>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 31247
keywords = []
message_count = 4.0
messages = ['300630', '300635', '300639', '300640']
nosy_count = 1.0
nosy_names = ['vstinner']
pr_nums = ['3166', '3168']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue31247'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
Using [PR 3138](https://github.com/python/cpython/pull/3138) of bpo-31234, I noticed that test_xmlrpc leaks dangling threads. Example:

haypo@selma$ ./python -m test -v test_xmlrpc --fail-env-changed -m test.test_xmlrpc.MultiPathServerTestCase.test_path1
== CPython 3.7.0a0 (heads/master:0267128, Aug 21 2017, 17:23:11) [GCC 6.4.1 20170727 (Red Hat 6.4.1-1)]
(...)
Warning -- threading_cleanup() detected 0 leaked threads (count: 0, dangling: 2)
test_xmlrpc failed (env changed)
(...)

It seems like xmlrpc error handling creates reference leaks. Attached PR fixes the issue.

**Comment by vstinner** [comment]
New changeset 84524454d0ba77d299741c47bd0f5841ac3f66ce by Victor Stinner in branch 'master':
bpo-31247: xmlrpc.server: break reference cycle (bpo-3166)
https://github.com/python/cpython/commit/84524454d0ba77d299741c47bd0f5841ac3f66ce

**Comment by vstinner** [comment]
New changeset 12a3e343e1667f91c0095a9a7dc15048cb620e41 by Victor Stinner in branch '3.6':
bpo-31247: xmlrpc.server: break reference cycle (bpo-3166) (bpo-3168)
https://github.com/python/cpython/commit/12a3e343e1667f91c0095a9a7dc15048cb620e41

**Comment by vstinner** [comment]
Bug fixed in Python 3.6 and 3.7 (master).

Python 2.7 doesn't seem to be affected by the bug, since Exception.\_\_traceback__ was only introduced in Python 3.
